dali/core/mm/mm_test_utils.h

```diff
@@ -0,0 +1,145 @@
+
+#include <gtest/gtest.h>
+#include <cstdlib>
+#include <unordered_map>
+#include <unordered_set>
+#include "dali/core/mm/memory_resource.h"
+#include "dali/core/mm/malloc_resource.h"
+#include "dali/core/mm/detail/util.h"
+
+namespace dali {
+namespace mm {
+
+template <bool owning, bool security_check,
+          typename Base, typename Upstream, typename... ExtraParams>
+class test_resource_wrapper : public Base {
+  using sentinel_t = uint32_t;
+  void *do_allocate(ExtraParams... extra, size_t bytes, size_t alignment) override {
+    if (simulate_oom_)
+      throw std::bad_alloc();
+    void *ret;
+    if (security_check) {
+      ret = upstream_->allocate(extra..., bytes + sizeof(sentinel_t), alignment);
+      detail::write_sentinel<sentinel_t>(ret, bytes);
+    } else {
+      ret = upstream_->allocate(extra..., bytes, alignment);
+    }
+    freed_.erase(ret);  // in case of address reuse
+    if (ret) {  // nullptr can be returned on success if 0 bytes were requested
+      auto it_ins = allocated_.insert({ret, { bytes, alignment }});
+      if (!it_ins.second) {
+        ASSERT_TRUE(!security_check && (bytes == 0 || it_ins.first->second.bytes == 0))
+          << "The allocator returned the same address twice for non-empty allocations.", ret;
+      }
+      it_ins.first->second = { bytes, alignment };
+    }
+    return ret;
+  }
+
+  void do_deallocate(ExtraParams... extra, void *ptr, size_t bytes, size_t alignment) override {
+    ASSERT_EQ(freed_.count(ptr), 0u) << "Double free of " << ptr;
+    auto it = allocated_.find(ptr);
+    ASSERT_NE(it, allocated_.end())
+      << "The address " << ptr << " was not allocated by this allocator";
+    ASSERT_EQ(it->second.bytes, bytes) << "Different size freed than allocated";
+    ASSERT_EQ(it->second.alignment, alignment) << "Diffrent alignment freed than allocated";
+    if (security_check) {
+      ASSERT_TRUE(detail::check_sentinel<sentinel_t>(ptr, bytes)) << "Memory corruption detected";
+    }
+    freed_.insert(ptr);
+    allocated_.erase(ptr);
+    if (security_check)
+      bytes += sizeof(sentinel_t);
+    upstream_->deallocate(extra..., ptr, bytes, alignment);
+  }
+
+  struct alloc_params {
+    size_t bytes = 0, alignment = 1;
+  };
+  std::unordered_map<void *, alloc_params> allocated_;
+  std::unordered_set<void *> freed_;
+  Upstream *upstream_ = nullptr;
+  bool simulate_oom_ = false;
+
+  bool do_is_equal(const Base &other) const noexcept override {
+    if (auto *oth = dynamic_cast<const test_resource_wrapper*>(&other))
+      return *upstream_ == *oth->upstream_;
+    else
+      return false;
+  }
+
+ public:
+  test_resource_wrapper() = default;
+  explicit test_resource_wrapper(Upstream *upstream) : upstream_(upstream) {}
+  test_resource_wrapper(const test_resource_wrapper &) = delete;
+  test_resource_wrapper(test_resource_wrapper &&) = delete;
+
+  ~test_resource_wrapper() {
+    reset();
+  }
+
+  void simulate_out_of_memory(bool enable) {
```

Comment:
Not used.

Reply:
I must have removed the test. I'll add one.

Reply:
Added.